docs/src/lucet-runtime/killswitch.md

```diff
@@ -0,0 +1,346 @@
+  - not an internal state but we will also discuss timeouts during a hostcall fault
+* `D -> E`
+
+### `A -> B` timeout
+
+If a timeout occurs between instance initialization (A) and the start of guest
+execution (B), there are two circumstances to consider: does the timeout occur
+before or after the Lucet instance has switched to `Domain::Guest`?
+
+#### Before switching to `Domain::Guest`
+
+The `KillSwitch` that fires acquires `terminable` and then locks
+`execution_domain` to determine timeout mechanism. This is before the instance
+has locked it in `enter_guest_region`, so it will acquire the lock, with a
+state of `Domain::Pending`. Seeing a pending instance, the `KillSwitch` will
+update it to `Domain::Cancelled` and release the lock, at which point the
+instance will acquire the lock, observe the instance is `Cancelled`, and return
+to the host without executing any guest code.
+
+#### After switching to `Domain::Guest`
+
+The `KillSwitch` that fires acquires `terminable` and then attempts to acquire
+a lock on `execution_domain` to determine timeout mechanism. Because the
+instance has already locked `execution_domain` to update it to `Domain::Guest`,
+this blocks until the instance releases the lock (and guest code is running).
+At this point, the instance is running guest code and it is safe for the
+`KillSwitch` to operate as if it were terminating any other guest code - with
+the same machinery as an instance in state `B` (a `SIGALRM`).
+
+### `A -> D` timeout
+
+This is a timeout while another `KillSwitch` has already fired, timing out a
+guest before exeuction. Because another `KillSwitch` must have fired for there
+to be a race, one of the two will acquire `terminable` and actually update
+`execution_domain`, while the other simply exits.
+
+### `B -> C` timeout
+
+If a timeout occurs during a transition from guest (B) to hostcall (C) code,
+there are two circumstances also contingent on if the instance has or has not
+yet switched to `Domain::Hostcall`.
+
+#### Before switching to `Domain::Hostcall`
+
+The `KillSwitch` that fires locks `execution_domain` and sees the execution
+domain is `Domain::Guest`. It then uses typical guest termination macinery and
+signals the guest. An important correctness subtlety here is that `KillSwitch`
+holds the `execution_domain` lock until the guest is terminated, so the guest
+cannot simultaneously proceed into hostcall code and receieve a stray
+`SIGALRM`.
+
+#### After switching to `Domain::Hostcall`
+
+The `KillSwitch` that fires acquires `terminable` and then attempts to acquire
+a lock on `execution_domain`. Because the instance is switching to or has
+switched to `Domain::Hostcall`, the `KillSwitch` will select the termination
+style for hostcalls. It will update the execution domain to
+`Domain::Terminated` and the instance will return when the hostcall exits and
+the `Terminated` domain is observed.
+
+### `B -> E` timeout, during guest fault or timeout
```

Comment:
This section is really hard for me to read.  I think it's related to the visual in which state E is really a lot of subtly different kind of states.  I'm also noticing this section has the word "if" a lot.  With that in mind, it might be beneficial to break this down even further.  Do you think it would be worthwhile to have separate sections for B -> E  (guest fault) and B -> E (timeout).  ?

Reply:
I've fiddled with the wording here and teased apart some of the sentences I overloaded. I hope this is clearer, though I'm concerned the actual `Timeout while handling a guest fault` section is still a bit dense. If you have a moment later, I'd appreciate your thoughts with a re-read.